src/lib.rs

```diff
@@ -155,118 +155,106 @@ pub fn extract_fragments(
-                            line
+                            "error parsing fragment identifier in fragment open symbol: {}
+                                     (used reserved character {})",
+                            line, c
                         )),
                     });
                 }
-
-                // If there no markers, append the line to the existing fragment.
-                fragment = fragment.map(|x| Fragment {
-                    body: if x.body.is_empty() {
-                        content.to_string()
-                    } else {
-                        x.body + "\n" + content
-                    },
-                    ..x
+            }
+        } else if let Some(col) = content.find(&symbols.fragment_close) {
+            if let Some(closed_fragment) = fragment_stack.pop() {
+                let trimmed_body = closed_fragment.body.trim_end_matches('\n').to_string();
+                if let Some(parent_fragment) = fragment_stack.last_mut() {
+                    // Special handling of "empty" fragments.
+                    if !trimmed_body.is_empty() {
+                        // Add the child fragments body to the parent fragment.
+                        parent_fragment.body.push_str(&trimmed_body);
+                        parent_fragment.body.push('\n');
+                    }
+                }
+                // Add the closed fragment to the results list
+                fragments.push(Fragment {
+                    body: trimmed_body,
+                    ..closed_fragment
+                });
+            } else {
+                return Err(FileError {
+                    err_type: ParseError::CloseBeforeOpen,
+                    filename: filename.to_owned(),
+                    line,
+                    col,
+                    message: Some("fragment close symbol found without an open symbol".to_string()),
+                });
+            }
+        } else if let Some(col) = content.find(&symbols.halt) {
+            // If the Fragment stack is not empty, we have an error as there is at least 1 open
+            // Fragment.
+            if !fragment_stack.is_empty() {
+                return Err(FileError {
+                    err_type: ParseError::HaltWhileOpen,
+                    filename: filename.to_owned(),
+                    line,
+                    col,
+                    message: Some(format!(
+                        "halt symbol found while a fragment was open: {}",
+                        line
+                    )),
                 });
             }
+            // Otherwise stop processing and break out.
+            break;
+        } else if let Some(fragment) = fragment_stack.last_mut() {
+            fragment.body.push_str(content);
+            fragment.body.push('\n');
         }
     }
 
+    if !fragment_stack.is_empty() {
+        return Err(FileError {
+            err_type: ParseError::DoubleOpen,
```

Comment:
**request:** Can you rename this error variant to reflect the changed semantics? Perhaps something like `UnclosedFragment`.

Reply:
Great pickup!